README.md

```diff
@@ -196,10 +212,245 @@ The following syntax shall be used to declare an array:
 ```YAML
 # Array of datatype uint32, by default size of the array is undefined
 datatype: uint32[]
+
 # Optional: specified number of elements in the array
 arraysize: 5
 ```
 
+Only single-dimensional arrays are supported.
+
+---------------
+# VALUE RANGE SPECIFICATION
+
+Methods, events, and properties can optionally specify a range of
+legal value that their `in`, `out`, `error` and property can take.
+
+A range is specified as a list of values that are legal for a given
+parameter or property to have.
+
+Range definitions can be used to verify that the contract of a method,
+event, or property is not violated by transmitting illegal values
+between caller and callee, or publisher and subscriber in the case of
+properties.
+
+If `range` is omitted, any value that can be carried by the given
+datatype is a legal value.
+
+A range is specified as a string with a boolean expression consisting
+of relational operators and operands set to the value being verified.
+
+The `$` operand is the value being verified.
+
+String values are double quoted (`"`). Double quotes within a string are expressed as two double quotes (`""`).
+
+
+`$ == 10 or $ == 20`
+
+`$ in (1,3,4,5)`
+
+`$ <= 20`
+
+`$ == "a value"`
+
+`$ == "a ""double-quoted"" value"`
+
+`$ != 1.2 and ($ < 1.0 or $ > 1.4)`
+
+If the expression evaluates to true, the value is considered legal
+
+## Using regular expressions
+
+A range expression can use regular expressions to match against value patterns.
+
+`regex("^the_.+_value$") or regex("^a_.+value$") or $ == "something_else"`
+
+Examples of legal values in this case would be `the_first_value`,
+`the_second_value`, `the_1234_value`, `something_else`.
+
+
+## Using sets
+
+A range expression can have a set:
+
+`$ in_set (1,2,5,10)`
+
+In this case `1`, `2`, `5`, and `10` are legal values while any other
+values are illegal.
+
+
+## Using intervals
+
+A range expression can have an interval:
+
+`$ in_interval (20,23)`
+
+In this case the legal values for the value are `20`, `21`, `22`, and `23`.
```

Comment:
Also valid for float? If so 20<=x<=23?

Reply:
To be complete it might need closed/open intervals.  either by name e.g. `closed_interval(...) ` or perhaps the [mathematical notation](https://www.teachoo.com/7017/1340/Interval-Notation/category/Intervals/)

Reply:
Added example for floats. 
Renaming to `closed_interval` is a separate discussion for a separate ticket.